src/ptvsd/adapter/ide.py

```diff
@@ -222,58 +260,101 @@ def launch_request(self, request):
 
-        pid = request("processId", int, optional=True)
-        if pid == ():
-            # When the adapter is spawned by the debug server, it is connected to the
-            # latter from the get go, and "host" and "port" in the "attach" request
-            # are actually the host and port on which the adapter itself was listening,
-            # so we can ignore those.
-            if self.server:
-                return
+        # There are four distinct possibilities here.
+        #
+        # If "processId" is specified, this is attach-by-PID. We need to inject the
+        # debug server into the designated process, and then wait until it connects
+        # back to us. Since the injected server can crash, there must be a timeout.
+        #
+        # If "subProcessId" is specified, this is attach to a known subprocess, likely
+        # in response to a "ptvsd_attach" event. If so, the debug server should be
+        # connected already, and thus the wait timeout is zero.
+        #
+        # If neither are specified, but "listen" is true, this is attach-by-socket
+        # with the server expected to connect to the adapter via ptvsd.attach(). There
+        # is no PID known in advance, so just wait until the first server connection
+        # indefinitely, with no timeout.
+        #
+        # If neither are specified, but "listen" is false, this is attach-by-socket
+        # in which the server has spawned the adapter via ptvsd.enable_attach(). There
+        # is no PID known to the IDE in advance, but the server connection should be
+        # there already, so the wait timeout is zero.
+        #
+        # In the last two cases, if there's more than one server connection already,
+        # this is a multiprocess re-attach. The IDE doesn't know the PID, so we just
+        # connect it to the oldest server connection that we have - in most cases, it
+        # will be the one for the root debuggee process, but if it has exited already,
+        # it will be some subprocess.
 
-            host = request("host", "127.0.0.1")
-            port = request("port", int)
-            if request("listen", False):
-                with self.accept_connection_from_server((host, port)):
-                    pass
-            else:
-                self.session.connect_to_server((host, port))
-        else:
-            if self.server:
+        pid = request("processId", int, optional=True)
+        sub_pid = request("subProcessId", int, optional=True)
+        if pid != ():
+            if sub_pid != ():
                 raise request.isnt_valid(
-                    '"attach" with "processId" cannot be serviced by adapter '
-                    "that is already associated with a debug server"
+                    '"processId" and "subProcessId" are mutually exclusive'
                 )
-
             ptvsd_args = request("ptvsdArgs", json.array(unicode))
```

Comment:
Unrelated to this PR. I think we should get rid of `ptvsdArgs`. Other than logging we don't really use it for anything. In case of logging we have a `--log-dir` anyway, that the IDE can pass to the adapter and server.

Reply:
It just needs to pass it to the adapter - the adapter already has the code to propagate it to the launcher, and from there to the debuggee.